doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,903 @@
+with are socket BIOs:
+
+  - The read-pollable FD is the FD of the read BIO.
+  - The write-pollable FD is the FD of the write BIO.
+
+However, things become more complex when we are dealing with memory-based BIOs
+such as `BIO_dgram_pair` which do not naturally correspond to any OS primitive
+which can be used for synchronisation, or when we are dealing with an
+application-provided custom BIO.
+
+### Pollable and Non-Pollable BIOs
+
+In order to accommodate these various cases, we draw a distinction between
+pollable and non-pollable BIOs.
+
+  - A pollable BIO is a BIO which can provide some kind of OS-level
+    synchronisation primitive, which can be used to determine when
+    the BIO might be able to do useful work once more.
+
+  - A non-pollable BIO has no naturally associated OS-level synchronisation
+    primitive, but its state only changes in response to calls made to it (or to
+    a related BIO, such as the other end of a pair).
+
+#### Supporting Pollable BIOs
+
+“OS-level synchronisation primitive” is deliberately vague. Most modern OSes use
+unified handle spaces (UNIX, Windows) though it is likely there are more obscure
+APIs on these platforms which have other handle spaces. However, this
+unification is not necessarily significant.
+
+For example, Windows sockets are kernel handles and thus like any other object
+they can be used with the generic Win32 `WaitForSingleObject()` API, but not in
+a useful manner; the generic readiness mechanism for WIndows handles is not
+plumbed in for socket handles, and so sockets are simply never considered ready
+for the purposes of this API, which will never return. Instead, the
+WinSock-specific `select()` call must be used. On the other hand, other kinds of
+synchronisation primitive like a Win32 Event must use `WaitForSingleObject()`.
+
+Thus while in theory most modern operating systems have unified handle spaces in
+practice there are substantial usage differences between different handle types.
+As such, an API to expose a synchronisation primitive should probably be of a
+tagged union design supporting possible variation.
+
+An SSL object will provide methods to retrieve a pollable OS-level
+synchronisation primitive which can be used to determine when the QUIC state
+machine can (potentially) do more work. In order to maintain the integrity of
+the BIO abstraction layer, an equivalent set of methods is added to the BIO API.
+
+Such a possible design could be:
+
+```c
+#define BIO_POLL_DESCRIPTOR_TYPE_NONE        0
+#define BIO_POLL_DESCRIPTOR_TYPE_SOCK_FD     1
+
+typedef struct bio_poll_descriptor_st {
+    int type;
+    union {
+        int fd;
+    } value;
+} BIO_POLL_DESCRIPTOR;
+
+int BIO_get_rpoll_descriptor(BIO *ssl, BIO_POLL_DESCRIPTOR *desc);
+int BIO_get_wpoll_descriptor(BIO *ssl, BIO_POLL_DESCRIPTOR *desc);
+
+int SSL_get_rpoll_descriptor(SSL *ssl, BIO_POLL_DESCRIPTOR *desc);
+int SSL_get_wpoll_descriptor(SSL *ssl, BIO_POLL_DESCRIPTOR *desc);
```

Comment:
The QUIC implementation needs something pollable it can use with select/etc. so that it can emulate blocking semantics. That means (if we want to preserve the BIO abstraction layer) a BIO_dgram at the very least must have some kind of method like this.

If we are defining this method for BIO_dgram I don't see much reason to avoid defining it for BIO_ssl as well.